libsql/src/hrana/connection.rs

```diff
@@ -72,162 +64,83 @@ where
-                            base_url,
-                        },
-                    );
-                }
-                None => Err(HranaError::StreamClosed(
-                    "Stream closed: server returned empty baton".into(),
-                ))?,
-            }
-        }
-
-        if response.results.is_empty() {
-            Err(HranaError::UnexpectedResponse(format!(
-                "Unexpected empty response from server: {:?}",
-                response.results
-            )))?;
-        }
-        if response.results.len() > 2 {
-            // One with actual results, one closing the stream
-            Err(HranaError::UnexpectedResponse(format!(
-                "Unexpected multiple responses from server: {:?}",
-                response.results
-            )))?;
-        }
-        match response.results.swap_remove(0) {
-            Response::Ok(StreamResponseOk {
-                response: StreamResponse::Execute(execute_result),
-            }) => Ok(execute_result.result),
-            Response::Ok(_) => Err(HranaError::UnexpectedResponse(format!(
+        match resp {
+            StreamResponse::Execute(resp) => Ok(resp.result),
+            other => Err(HranaError::UnexpectedResponse(format!(
                 "Unexpected response from server: {:?}",
-                response.results
+                other
             ))),
-            Response::Error(e) => Err(HranaError::StreamError(e)),
         }
     }
 
-    async fn _close_stream_for(&self, tx_id: u64) -> Result<()> {
-        let client = self.client();
-        let cookie = client
-            .cookies
-            .read()
-            .unwrap()
-            .get(&tx_id)
-            .cloned()
-            .unwrap_or_default();
-        let msg = ClientMsg {
-            baton: cookie.baton,
-            requests: vec![StreamRequest::Close],
-        };
-        let url = cookie
-            .base_url
-            .unwrap_or_else(|| client.url_for_queries.clone());
-        let body = serde_json::to_string(&msg).map_err(HranaError::Json)?;
-        client
-            .inner
-            .http_send(url, client.auth.clone(), body)
-            .await
-            .ok();
-        client.cookies.write().unwrap().remove(&tx_id);
-        Ok(())
-    }
-
     pub fn prepare(&self, sql: &str) -> Statement<T> {
-        Statement {
-            client: self.clone(),
-            inner: Stmt::new(sql, true),
-        }
+        Statement::from_connection(self.clone(), sql.to_string(), true)
     }
 }
 
-impl<T> Clone for HttpConnection<T> {
+impl<T> Clone for HttpConnection<T>
+where
+    T: for<'a> HttpSend<'a>,
+{
     fn clone(&self) -> Self {
         HttpConnection(self.0.clone())
     }
 }
+
+pub(crate) enum CommitBehavior {
+    Commit,
+    Rollback,
+}
+
+pub(super) fn stmts_to_batch(protocol_v3: bool, stmts: impl IntoIterator<Item = Stmt>) -> Batch {
+    let mut batch = Batch::new();
+    let mut step = -1;
+    for stmt in stmts.into_iter() {
```

Comment:
could use https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.enumerate maybe here to do the counter for you so you don't need to do that logic manually.